security/advancedtls/crl.go

```diff
@@ -0,0 +1,489 @@
+	UNREVOKED
+	// REVOKED means we found the CRL and the cert is revoked
+	REVOKED
+)
+
+func (s RevocationStatus) String() string {
+	return [...]string{"UNDETERMINED", "UNREVOKED", "REVOKED"}[s]
+}
+
+// certificateListExt contains a pkix.CertificateList and parsed
+// extensions that aren't provided by the golang CRL parser
+type certificateListExt struct {
+	CertList *pkix.CertificateList
+	// RFC5280, 5.2.1, all conforming CRLs must have a AKID with the ID method
+	AuthorityKeyID []byte
+}
+
+const tagDirectoryName = 4
+
+var (
+	// RFC5280, 5.2.4 id-ce-deltaCRLIndicator OBJECT IDENTIFIER ::= { id-ce 27 }
+	oidDeltaCRLIndicator = asn1.ObjectIdentifier{2, 5, 29, 27}
+	// RFC5280, 5.2.5 id-ce-issuingDistributionPoint OBJECT IDENTIFIER ::= { id-ce 28 }
+	oidIssuingDistributionPoint = asn1.ObjectIdentifier{2, 5, 29, 28}
+	// RFC5280, 5.3.3 id-ce-certificateIssuer   OBJECT IDENTIFIER ::= { id-ce 29 }
+	oidCertificateIssuer = asn1.ObjectIdentifier{2, 5, 29, 29}
+	// RFC5290, 4.2.1.1 id-ce-authorityKeyIdentifier OBJECT IDENTIFIER ::=  { id-ce 35 }
+	oidAuthorityKeyIdentifier = asn1.ObjectIdentifier{2, 5, 29, 35}
+)
+
+// X509NameHash implements the OpenSSL X509_NAME_hash function for hashed directory lookups.
+func X509NameHash(r pkix.RDNSequence) string {
+	var canonBytes []byte
+	// First canonicalize all the strings
+	for _, rdnSet := range r {
+		for i, rdn := range rdnSet {
+			value, ok := rdn.Value.(string)
+			if !ok {
+				continue
+			}
+			// OpenSSL trims all whitespace, does a tolower, and removes extra spaces between words
+			// Implemented in x509_name_canon in OpenSSL
+			canonStr := strings.Join(strings.Fields(
+				strings.TrimSpace(strings.ToLower(value))), " ")
+			// Then it changes everything to UTF8 strings
+			rdnSet[i].Value = asn1.RawValue{Tag: asn1.TagUTF8String, Bytes: []byte(canonStr)}
+
+		}
+	}
+
+	// Finally, OpenSSL drops the initial sequence tag
+	// So, we marshal all the RDNs separately instead of as a group
+	for _, canonRdn := range r {
+		b, err := asn1.Marshal(canonRdn)
+		if err != nil {
+			log.Print(err)
+		}
+		canonBytes = append(canonBytes, b...)
+	}
+
+	issuerHash := sha1.Sum(canonBytes)
+	// Openssl takes the first 4 bytes and encodes them as a little endian
+	// uint32 and then uses the hex to make the file name
+	// In C++, this would be:
+	// (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
+	// ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)
+	// ) & 0xffffffffL;
+	fileHash := binary.LittleEndian.Uint32(issuerHash[0:4])
+	return fmt.Sprintf("%08x", fileHash)
+}
+
+// CheckRevocation checks the connection for revoked certificates based on RFC5280
+func CheckRevocation(conn tls.ConnectionState, cfg Config) error {
+	return CheckChainRevocation(conn.VerifiedChains, cfg)
+}
+
+// CheckChainRevocation checks the verified certificate chain
+// for revoked certificates based on RFC5280
+func CheckChainRevocation(verifiedChains [][]*x509.Certificate, cfg Config) error {
```

Comment:
Is the revocation supposed to happen after the cert verification, or before the verification?
From the API defined here, it seems the revocation shall happen after the cert verification, after we already obtained the verified chain? If that's the case, I think we don't even need to do the whole tricks in `advancedtls` to first disable the default cert verification and then re-write the verification again - I made that suggestion based on the assumption that the revocation should happen before the cert verification.  
Just wanted to make sure what's the correct way to handle revocations, because I am not quite familiar with it:)

Reply:
According to [RFC4158](https://datatracker.ietf.org/doc/html/rfc4158), path building should take revocation into account. The real answer to this question is that it depends on how golang does path building. If golang stops at the first valid path to a trust anchor, then passes it to this library and it's revoked, but there was another path that could have worked, the handshake will fail, when there was a path that would have worked. This implementation will create something similar to this [post](https://medium.com/@sleevi_/path-building-vs-path-verifying-the-chain-of-pain-9fbab861d7d6)